src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -531,390 +524,294 @@ func setHelmLabelsAndAnnotations(object client.Object, rp *v1alpha2.Redpanda) {
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            rp.GetHelmReleaseName(),
-			Namespace:       rp.Namespace,
-			OwnerReferences: []metav1.OwnerReference{rp.OwnerShipRefObj()},
-		},
-		Spec: helmv2beta2.HelmReleaseSpec{
-			Chart: helmv2beta2.HelmChartTemplate{
-				Spec: helmv2beta2.HelmChartTemplateSpec{
-					Chart:    "redpanda",
-					Version:  rp.Spec.ChartRef.ChartVersion,
-					Interval: &metav1.Duration{Duration: 1 * time.Minute},
-					SourceRef: helmv2beta2.CrossNamespaceObjectReference{
-						Kind:      "HelmRepository",
-						Name:      rp.GetHelmRepositoryName(),
-						Namespace: rp.Namespace,
-					},
-				},
-			},
-			Values:   values,
-			Interval: metav1.Duration{Duration: 30 * time.Second},
-			Timeout:  timeout,
-			Upgrade:  upgrade,
-		},
-	}, nil
-}
-
-func (r *RedpandaReconciler) createHelmRepositoryFromTemplate(rp *v1alpha2.Redpanda) *sourcev1.HelmRepository {
-	return &sourcev1.HelmRepository{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            rp.GetHelmRepositoryName(),
-			Namespace:       rp.Namespace,
-			OwnerReferences: []metav1.OwnerReference{rp.OwnerShipRefObj()},
-		},
-		Spec: sourcev1.HelmRepositorySpec{
-			Interval: metav1.Duration{Duration: 30 * time.Second},
-			URL:      v1alpha2.RedpandaChartRepository,
-		},
+	patch := client.MergeFrom(rp.DeepCopy())
+	controllerutil.RemoveFinalizer(rp, FinalizerKey)
+	if err := r.Patch(ctx, rp, patch); err != nil {
+		log.Error(err, "finalizer not removed")
 	}
-}
 
-func (r *RedpandaReconciler) patchRedpandaStatus(ctx context.Context, rp *v1alpha2.Redpanda) error {
-	key := client.ObjectKeyFromObject(rp)
-	latest := &v1alpha2.Redpanda{}
-	if err := r.Client.Get(ctx, key, latest); err != nil {
-		return err
-	}
-	return r.Client.Status().Patch(ctx, rp, client.MergeFrom(latest))
+	return ctrl.Result{}, nil
 }
 
-// event emits a Kubernetes event and forwards the event to notification controller if configured.
-func (r *RedpandaReconciler) event(rp *v1alpha2.Redpanda, revision, severity, msg string) {
-	var metaData map[string]string
-	if revision != "" {
-		metaData = map[string]string{v2.GroupVersion.Group + revisionPath: revision}
+func (r *RedpandaReconciler) reconcileStatus(ctx context.Context, rp *v1alpha2.Redpanda) error {
+	log := ctrl.LoggerFrom(ctx).WithName("RedpandaReconciler.reconcileStatus")
+
+	rp.Status.ObservedGeneration = rp.Generation
+
+	dot, err := rp.GetDot()
+	if err != nil {
+		log.Error(err, "failed to get dot")
+	}
+	sel := redpanda.StatefulSetPodLabelsSelector(dot)
+
+	// Get redpanda Pods readiness
+	var pods corev1.PodList
+	err = r.List(ctx, &pods, &client.ListOptions{
+		Namespace:     rp.Namespace,
+		LabelSelector: labels.SelectorFromValidatedSet(sel),
+		FieldSelector: fields.SelectorFromSet(fields.Set{
+			"status.phase": string(corev1.PodRunning),
+		}),
+	})
+	if err != nil {
+		log.Error(err, "failed to list pods")
+	}
+	rpPodsNum := len(pods.Items)
+
+	// TODO (Rafal) Make it a function within helm chart
```

Comment:
We decided we were not going to run the jobs within the post-flux operator, right?

Reply:
I was kind of wondering about that here:

1. How necessary are the post upgrade jobs? I'm not familiar enough with them to know _exactly_ what they do.
2. Is it possible just to implement the jobs directly in controller code that fires wnen a CRD changes or as part of some sort of init container for a node? (when do we actually need to run the code that is currently in the post upgrade job?)

Reply:
Yes, I just wanted to correctly handle Status report, so my test is validating:
* Operator can reconcile Redpanda CR without generating to many generation
* Operator report Status "Success" when all resources are reconciled

I would be more than happy to use go code to configure Redpanda cluster.

Reply:
> How necessary are the post upgrade jobs? I'm not familiar enough with them to know exactly what they do.

post upgrade or post-install jobs:
* set's Redpanda license as you can not bootstrap Redpanda with license as far as I know you need to pass it via Admin API rest'ish HTTP call.
* Reconfigures Redpanda cluster properties that does not require restart.
* Resets Schema Registry service as super user change might affect connection between Schema registry and Redpanda Kafka API.

> Is it possible just to implement the jobs directly in controller code that fires wnen a CRD changes or as part of some sort of init container for a node? (when do we actually need to run the code that is currently in the post upgrade job?)

Yes we have task for that. https://redpandadata.atlassian.net/browse/K8S-328

Reply:
The job itself was primarily futzing with getting variables into the correct place. This is the go version of both jobs: https://github.com/redpanda-data/redpanda-operator/pull/231 It'd be easier to just copy paste it honestly.

Reply:
> It'd be easier to just copy paste it honestly.

Agree, I will adjust that.